### ISSUE TITLE
Uploading /wp-includes/ JS and CSS to s3

### ISSUE THREAD

**Issue by Lunteer** [humanmade/S3-Uploads#276]
More of a question/feature request than an issue (I am in the wrong place?) 

I am looking to offload as much of the load on my server to s3 as possible. This plugin works great for the wp-content folder. Anyway to get this to work with the /wp-includes/ folder? It seems that the php files need to be on the server but would love to move the JS and CSS to S3.

Bonus question: Anyway to get the /wp-includes/ JS and CSS files to be in their own bucket, so I can setup a new CloudFront subdomain? Does it need to be it's own bucket or can I setup CloudFront to only look at a specific folder in the bucket?

To go one step further, can this also be done with JS and CSS under /wp-admin/?

FWIW, it seems like the WP Super Cache plugin does something similar with the basic CDN settings. (no upload to s3 but rewrites for CDN. I would want the upload and URL rewrite).

Thanks and sorry in advance if this is the wrong location.

**Comment by rmccue** [comment]
> More of a question/feature request than an issue (I am in the wrong place?)

This is the right place. :)

It sounds to me like what you want to do is not so much offload to S3, but more offload to CloudFront. Is there any reason I'm missing why you'd need S3 specifically for this?

If all you want is to reduce the server load, then I'd recommend setting up CloudFront in front of WordPress and setting up custom cache behaviours for `wp-includes` in the distribution. The same can be done for `wp-admin/js`, etc, fairly easily.

You can also set up your underlying server to send correct cache headers for static files, then let CloudFront use those headers to determine its behaviour (i.e. "Use Origin Cache Headers").

**Comment by Lunteer** [comment]
@rmccue thanks for the idea. I do not necessarily need it to be on a s3 (although doing so will mean Cloudfront never hits my server for css or js files), that is just what I have been reading online (Thanks for broadening my horizon!!!) and seems to be inline with the same behavior as the `wp-content` folder.

"fairly easy" is relative and I know that even if I use Cloudfront Origin Pull I still need to rewrite URLs and that is where I got flat faced. Not interested in heavy plugins like WP Super Cache or W3 Total Cache, which is why I am here 😄 .

So in short (I guess) and with the insights you have given me so far and whether I decide to use s3 or Origin Pull. Can I use S3 Uploads to rewrite URLs (preferably with my own subdomain like wp.example.com) for `wp-includes` and `wp-admin/js`?

**Comment by Lunteer** [comment]
Additional question along the same lines. Is it possible to move the `themes`, `mu-plugins` (almost the entire `wp-content`) onto s3 or load with Cloudfront?

**Comment by Lunteer** [comment]
Partial answer to this problem:

For those who want to know what is still missing:
Files are on S3 and using Cloudfront for CDN 🍾 . Need to write code within Wordpress to change the domain URL when calling JS and CSS files within wp-admin and wp-includes folders. So instead of `www.example.com/wp-includes/css/dashicons.min.css` it points to `wp-core.example.com/wp-includes/css/some-css-file.css`. We are using a constant in our wp-config.php file to define the domain but need to workout the filters.

This process is heavy but with some bash scripts fairly easy to automate.

### Copy JS/CSS to temp folder
You first want to create a folder to where you will copy the JS/CSS files (you need this because AWS CLI uses sync and you cannot select file types - at least I did not find a way.

`rsync -zrvm --include "*/"  --include="*.js" --include="*.css" --exclude="*" /path/to/folder/wp-admin/ /path/to/tmp/folder/wp-admin/`

Repeat for all the folder (sample list):

- wp-admin
- wp-includes
- wp-content/themes
- wp-content/wp-plugins
- wp-content/mu-plugins

when doing the wp-content folders you do not want to do the `wp-content/uplaods` folder as this is what S3-Uploads does 😃 .

###  Upload to S3
Next step is to upload to S3. You can do this in bulk or on each update only upload that specific directory

_We decided to go with a different S3 bucket and CDN, instead of combining with the `/uploads/` folder. I know back in the day people pushed for multiple calls to help speed up sites. Things have changed and it seemed easier to manager for us._

To upload in bulk:

`aws s3 sync ~/tmp-folder-holding-JS-CSS/ s3://bucket-name/`

To upload just one folder:

`aws s3 sync ~/tmp-folder-holding-JS-CSS/wp-admin/ s3://bucket-name/wp-admin/`

**Important Note**
Be careful using the --delete flag when uploading just one folder. If done wrong it can delete all the other folders 😨 

### Changing the URLs

Last part as written above is actually rewriting the URLs so they point to the new CDN. Filtering seems to be the issue and once we finish that part we will post it here. (any recommendations are welcome.

### S3 as a File System

In all the research we did for this concept we came across, https://cloud.netapp.com/blog/amazon-s3-as-a-file-system which the next level up. We decided against the idea for now because of the latency. In truth that is not such an issue because we are using Opcache so that would store all the PHP files so not using the file system as much.

Any thoughts/ideas/criticisms are more than welcome

**Comment by prafulbagai** [comment]
To change content URL, put `define('WP_CONTENT_URL', 'http://static.yourdomain.com');` in your config.

**Comment by Lunteer** [comment]
@prafulbagai Mind Blown 🤯. Thank you, Thanks You!

Just a quick questions... 
Does this constant, WP_CONTENT_URL, have to be declared after the s3-uploads constants or the order makes no difference?

I ask because the /uploads/ folder was not affected and we declared it after the s3-uploads constants. If the s3-uploads plugin was not installed would it have also changed the /uploads/ folder?

Thanks Again

**Comment by abhishek-pokhriyal** [comment]
Hey @Lunteer, it looks like I've similar kind of issues. Can you help me by answering some of the questions?

> _I'm using CloudFront as my CDN, with S3 as my data source, for my WordPress site._ Also, I'm using WP Rocket plugin for CDN integration and WP Offload Media Lite for offloading media uploads as they are added.


1. Till now I've uploaded the `wp-content/uploads` folder into my S3 bucket. Now, I want to make it work for my `js` and `css` assets. Is it a good idea to upload my complete `wp-content` folder for that purpose?

1. Currently, the images that are added to the media library automatically goes to the S3 bucket (thanks to WP Offload Media Lite plugin). Is there a way that I can do the same for `js` and `css` files?

Thank you

**Comment by Lunteer** [comment]
@abhishek-kaushik 

1. I upload `css` in the /uploads/ folder because the entire folder is on s3 and on the same CDN. If you didn't it would return a 403 error. My issue was with `/wp-content/themes/`, `/wp-content/plugins/` folders where I wanted those `js` and `css` files also in an s3 bucket and on a CDN. 

To answer your question directly, there are a ton of `.php` files in the `wp-content` folder that is not used or needed in the s3 bucket. Can you push it there, sure, but it will not be used and also not cost you anything. I personally created scripts to only push the files needed into s3, leaving the php files locally.

2. I use this plugin, which is the thread of the git repository here :), to automatically push all files in the uploads to the s3 bucket. For all other files I want in s3, I wrote bash scripts in linux to do the pushing to s3. I only have the scripts run when I update plugins, themes and/or WordPress. Two things to note (1) if you are using CDN for anything else besides images you might need to clear the CDN cache and (2) I pushed basically all files for WordPress (except PHP and some other files) to an s3 bucket and server over CDN. Takes a huge amount of load of servers and allows for greater scale. (Thing even the `/wp-admin/` `js` and `css` run off a CDN).

Hope this helps!

**Comment by abhishek-pokhriyal** [comment]
> To answer your question directly, there are a ton of .php files in the wp-content folder that is not used or needed in the s3 bucket.

You got my exact concern! Thank you.

> I wrote bash scripts in linux to do the pushing to s3.

Gotcha 👍 

Thank you for responding back. Those were really helpful.

PS: you've replied to some other GitHub user.
